src/time-to-interactive/TTIBoomerang.ts

```diff
@@ -0,0 +1,331 @@
+
+
+We do not record PageBusy metrics for now as there were some performance issues observed. 
+As such TTI can only be computed for now for browsers that support Long Tasks. 
+*/
+
+export class TTIBoomerang {
+    private fcpTime: number | null = null;
+    private lcpTime: number | null = null;
+    private domContentLoadedEventEnd: number | null = null;
+    private visuallyReadyTimestamp: number | null = null;
+    private ttiTracker: any = {};
+    private ttiResolved = false;
+
+    private fcpSupported = false;
+    private lcpSupported = false;
+    private fpsSupported = false; // Check for support for requestAnimationFrame
+
+    private COLLECTION_PERIOD = 100;
+    private REQUIRED_OK_INTERVALS = 5;
+    private CHECK_PERIOD = 1000; // Perform another check for TTI/Visually ready
+    private VISUALLY_READY_RESOLVE_TIMEOUT = 10000;
+    private TTI_RESOLVE_TIMEOUT = 10000;
+
+    private FPS_THRESHOLD = 20 / (1000 / this.COLLECTION_PERIOD);
+    private LONG_TASK_THRESHOLD = 0;
+
+    // Needs to return a Promise
+    computeTimeToInteractive(): Promise<number> {
+        this.initListeners();
+        return new Promise<number>((resolve, reject) => {
+            this.checkForVisualReady().then((result) => {
+                this.computeTTI(result).then((result) => {
+                    resolve(result);
+                });
+            });
+        });
+    }
+
+    checkForVisualReady(): Promise<number> {
+        // Check if visually ready
+        return new Promise<number>((resolve, reject) => {
+            let timeIntervals = 0;
+            const visuallyReadyInterval: any = setInterval(() => {
+                // Check if all Visually ready timestamps are available
+                let visuallyReady = true;
+
+                if (this.fcpSupported) {
+                    if (!this.fcpTime) {
+                        visuallyReady = false;
+                    }
+                }
+                if (this.lcpSupported) {
+                    if (!this.lcpTime) {
+                        visuallyReady = false;
+                    }
+                }
+
+                if (!this.domContentLoadedEventEnd) {
+                    visuallyReady = false;
+                }
+
+                if (
+                    visuallyReady ||
+                    timeIntervals * this.CHECK_PERIOD >
+                        this.VISUALLY_READY_RESOLVE_TIMEOUT
+                ) {
+                    // If timed out, check if any visually ready signals are present, else can't find tti
+                    if (
+                        timeIntervals * this.CHECK_PERIOD >
+                        this.VISUALLY_READY_RESOLVE_TIMEOUT
+                    ) {
+                        if (
+                            !this.lcpTime &&
+                            !this.domContentLoadedEventEnd &&
+                            !this.fcpTime
```

Comment:
question: Isn't this always false, since `visualReady` must be true here? Or am I not following the logic correctly?